add cmd rotate-encryption-key

kidd opened the pull request:
This PR adds `rotate-encryption-key` command line to decrypt and reencrypt data with a new key.

The command is intended to be used on a stopped mb instance. After the command succeeds, metabase should be restarted with the new `MB_ENCRYPTION_SECRET_KEY`.

We don't allow to rotate to an empty key from the cli, although this functionality is used internally by `dump-to-h2 --dump-plaintext`.

TODO:
- [x] make sure dump-to-h2 --dump-plaintext notifies differently when it succeeds and when it doesn't

kidd committed:
add cmd rotate-encryption-key

kidd commented on src/metabase/cmd/rotate_encryption_key.clj:
```diff
@@ -0,0 +1,30 @@
+                       :metabase_database
+                       {:details (encrypt/maybe-encrypt hashed-key (json/encode details))}
+                       ["id = ?" id])))
+      true
```
This is super arbitrary but I think it is nice to provide some extra info to the caller about "how did it go". Is it ok?

kidd commented on src/metabase/models/setting.clj:
```diff
@@ -55,7 +55,8 @@
   (merge models/IModelDefaults
-         {:types (constantly {:value :encrypted-text})}))
+         {:types (constantly {:value :encrypted-text})
+          :primary-key (constantly :key)}))
```
Not related to this PR, but I think it's just missing. 
Is it ok to put it here?

camsaul replied:
yeah 👍

kidd committed:
enable rotate secret key to nil in order to decrypt

* Allow to  decrypt a db to plaintext.
* Add more checks for key lenght

kidd committed:
check database.details decryption when rotating-keys

rollback key rotation if a decryption key is not valid.

kidd committed:
make sure we start from clean encryption key

kidd committed:
lowercase database

kidd committed:
fix(mysql) add table prefix to column names in jdbc queries

kidd committed:
test(rotate-keys) test databases details rotation

kidd committed:
refactor dump-to-h2 --dump-plaintext

--dump-plaintext uses now `rotate-keys!`.

kidd committed:
rename function to the command name

kidd commented on src/metabase/cmd/rotate_encryption_key.clj:
```diff
@@ -0,0 +1,35 @@
+(defn rotate-encryption-key!
+  "Rotate the current configured db using the current MB_ENCRYPTION_SECRET_KEY env var and `to-key` argument."
+  [to-key]
+  (mdb/setup-db!)
```
where should I put this call? cmd.clj doesn't have calls to this function, but it doesn't feel right here either.  Or, maybe it's not needed? I'll recheck

kidd commented on test/metabase/cmd/rotate_encryption_key_test.clj:
```diff
@@ -0,0 +1,102 @@
+
+(defn- raw-value [key]
+  (:value (first (jdbc/query mdb.connection/*jdbc-spec*
+                             ["select value from setting where setting.key=?;" key]))))
```
mysql required to add the tablename `setting` in the filter. Is it really needed? or maybe there's a config to not require it?

flamber replied:
Could it be because `value` and `key` are reserved words and would require backtick enclosure like `` `value` `` ?
https://dev.mysql.com/doc/refman/8.0/en/keywords.html

kidd replied:
ah, that might be it. Thanks!

kidd committed:
explicit check of return value

dump-to-h2 should return a positive number to indicate failure to dump
correctly. rotate-encryption-key cmd uses now the same approach.

kidd committed:
improved code for logging

kidd commented on src/metabase/cmd.clj:
```diff
@@ -52,8 +52,9 @@
         return-code    ((resolve 'metabase.cmd.dump-to-h2/dump-to-h2!) h2-filename options)]
-    (when (pos-int? return-code)
-      (system-exit! return-code))))
+    (if (pos-int? return-code)
```
I'm trying to keep the same interface of "if return positive int" because it was already there, but, it feels to me this case won't be reached every, because `dump-to-h2!` either ends ok, or throws an exception (copy/copy! throws them).

Am I missing something? if the pos-int? interface is not used at all and we rely on the exception being thrown or not, I can simplify this a bit more

robdaemon replied:
I think that's the issue - the exception being thrown leads down a path where there is no `(system-exit! 1)` or anything. I think the main command dispatcher may need a try...catch to handle exceptional conditions

kidd committed:
add tests for rotate errors

kidd commented:
@robdaemon  wrt `--dump-plaintext`, this is the testruns I did (most of them tested in the tests, but to be extra sure)
| exit code | cmd                                                                                                |
|         0 | lein run dump-to-h2 /tmp/d.db --dump-plaintext                                                     |
|         1 | MB_ENCRYPTION_SECRET_KEY=a$MB_ENCRYPTION_SECRET_KEY lein run dump-to-h2 /tmp/d.db --dump-plaintext |
|         1 | MB_ENCRYPTION_SECRET_KEY='' lein run dump-to-h2 /tmp/d.db --dump-plaintext                         |
|         1 | MB_ENCRYPTION_SECRET_KEY= lein run dump-to-h2 /tmp/d.db --dump-plaintext                           |
|         1 | unset MB_ENCRYPTION_SECRET_KEY;  lein run dump-to-h2 /tmp/d.db --dump-plaintext                    |

camsaul commented on src/metabase/cmd.clj:
```diff
@@ -152,6 +153,14 @@
+  [new-key]
+  (classloader/require 'metabase.cmd.rotate-encryption-key)
+  (let [return-code ((resolve 'metabase.cmd.rotate-encryption-key/rotate-encryption-key!) new-key)]
+    (when (not return-code)
+      (system-exit! 1))))
```
You should use `when-not` here instead of `when (not ...)`

camsaul commented on src/metabase/cmd/dump_to_h2.clj:
```diff
@@ -34,6 +36,10 @@
+                 mdb.conn/*jdbc-spec* h2-jdbc-spec
+                 db/*db-connection* h2-jdbc-spec
+                 db/*quoting-style* :h2]
+         (when (not (rotate-encryption/rotate-encryption-key! nil))
+           1))))))
```
This is kind of weird to return a number like this. Is this supposed to be the exist code? Like Rob suggested, you should just throw exceptions if stuff goes wrong and then have the top level of the command catch them and then print them and exit with a nonzero status

kidd replied:
aha, yes, it felt strange when I was doing it already. I'll change it to exception being thrown.

camsaul commented on src/metabase/cmd/rotate_encryption_key.clj:
```diff
@@ -0,0 +1,35 @@
+          (when (encrypt/possibly-encrypted-string? details)
+            (throw  (Exception. "MB_ENCRYPTION_SECRET_KEY does not correcty decrypt files")))))
+      true)
+    (catch Throwable e (println "MB_ENCRYPTION_SECRET_KEY does not correcty decrypt the existing data"))))
```
Instead of println you should use `log/error`, `log/debug`, `log/info`, etc. in your commands. It lets us turn off the logging in tests for example

camsaul reviewed (approved):
Generally this looks good to me other than the way you're doing return codes as Rob mentioned. Just catch exceptions at the top level and return non-zero, otherwise it should call `(System/exit 0)`. Otherwise it is liable to hang when it's done while it's waiting for various threadpools to finish (which may never happen)

kidd committed:
change dump-to-h2 return value interface

dump-to-h2! and rotate-encryption-key! do throw an exception if they
couldn't complete successfully.

The cmd that call them catch the exception and return 1 to the shell.

kidd committed:
use log/error to show errors in cmds
